### PR TITLE
[coordinator] Add multi-process config for on-prem process scaling

### PR DESCRIPTION
**Pull request by robskillington**
**What this PR does / why we need it**:

This adds process scaling for on-premise workloads by allowing them to deploy many smaller processes that can also then be pinned to NUMA nodes if desired. This is essentially the same as placing a L4 TCP load balancer on the local machine and allowing multiple processes to serve the request. 

A future change could use a L7 load balancer in the parent process and distribute requests to unix domain socket listen servers per-process.

This should be considered experimental and only used when all other options have been exhausted (such as multiple-process supervision/configuration with your favorite configuration tool and/or compute framework such as Kubernetes/Nomad).

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/cmd/services/m3query/config/config.go [hunk 1]
```diff
@@ -543,3 +546,18 @@ func TagOptionsFromConfig(cfg TagOptionsConfiguration) (models.TagOptions, error
 type ExperimentalAPIConfiguration struct {
 	Enabled bool `yaml:"enabled"`
 }
+
+// MultiProcessConfiguration is the multi-process configuration which
+// allows running multiple sub-processes of an instance reusing the
+// same listen ports.
+type MultiProcessConfiguration struct {
+	// Enabled is whether ot enable multi-process execution.
```
Comment (arnikola) [inline src/cmd/services/m3query/config/config.go:554]: nit; to

File: src/x/net/listen.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+// Listen creates a new listener based on options.
+func (o ListenerOptions) Listen(
+	protocol, address string,
+) (gonet.Listener, error) {
+	if o.reusePort {
+		if protocol == "tcp" {
+			protocol = "tcp4"
+		}
+		fmt.Printf("!! using REUSE port for address=%s\n", address)
```
Comment (arnikola) [inline src/x/net/listen.go:73]: nit delete

File: src/query/server/query.go [hunk 3]
```diff
@@ -172,6 +184,109 @@ func Run(runOpts RunOptions) {
 
 	xconfig.WarnOnDeprecation(cfg, logger)
 
+	if cfg.MultiProcess.Enabled {
+		multiProcessInstance := os.Getenv(multiProcessInstanceEnvVar)
+		if multiProcessInstance == "" {
+			logger = logger.With(zap.String("processID", "parent"))
+			count := cfg.MultiProcess.Count
+			if count == 0 {
```
Comment (arnikola) [inline src/query/server/query.go:193]: nit: `count < 1` to be safe? Can log the value if it's ignored due to being non zero

File: src/cmd/services/m3query/config/config.go [hunk 4]
```diff
@@ -543,3 +546,18 @@ func TagOptionsFromConfig(cfg TagOptionsConfiguration) (models.TagOptions, error
...
+	Enabled bool `yaml:"enabled"`
+	// Count is the number of sub-processes to run, leave zero
+	// to auto-detect based on number of CPUs.
+	Count int `yaml:"count"`
+	// PerCPU is the factor of processes to run per CPU, leave
+	// zero to use the default of 0.5 per CPU (i.e. one process for
+	// every two CPUs).
+	PerCPU float64 `yaml:"perCPU"`
```
Comment (arnikola) [inline src/cmd/services/m3query/config/config.go:562]: Add config verification here? Probably still a good practice to keep it in other places, but may as well catch it at the source?

```suggestion
	Count int `yaml:"count" validate:"min=0"`
	// PerCPU is the factor of processes to run per CPU, leave
	// zero to use the default of 0.5 per CPU (i.e. one process for
	// every two CPUs).
	PerCPU float64 `yaml:"perCPU" validate:"min=0.0, max=0.0"`
```

File: src/query/server/query.go [hunk 5]
```diff
@@ -172,6 +184,109 @@ func Run(runOpts RunOptions) {
 
 	xconfig.WarnOnDeprecation(cfg, logger)
 
+	if cfg.MultiProcess.Enabled {
+		multiProcessInstance := os.Getenv(multiProcessInstanceEnvVar)
+		if multiProcessInstance == "" {
+			logger = logger.With(zap.String("processID", "parent"))
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/query/server/query.go:191]: So this is the parent spawning process, maybe I'll call it instance "0" and the rest can get a "1" indexed processID (i.e. starting at "1").

File: src/query/server/query.go [hunk 6]
```diff
@@ -172,6 +184,109 @@ func Run(runOpts RunOptions) {
...
+			count := cfg.MultiProcess.Count
+			if count == 0 {
+				perCPU := defaultPerCPUMultiProcess
+				if v := cfg.MultiProcess.PerCPU; v > 0 {
```
Comment (arnikola) [inline src/query/server/query.go:195]: Upper bound of 1?

File: src/cmd/services/m3query/config/config.go (hunk 1 above)
Comment (robskillington) [reply src/cmd/services/m3query/config/config.go:554]: TY

File: src/x/net/listen.go (hunk 2 above)
Comment (robskillington) [reply src/x/net/listen.go:73]: Ah yes, TY for the catch

File: src/query/server/query.go [hunk 7]
```diff
@@ -172,6 +184,109 @@ func Run(runOpts RunOptions) {
...
+			)
+			for i := 0; i < count; i++ {
+				i := i
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+
+					newEnv := []string{
+						fmt.Sprintf("%s=%s",
+							multiProcessInstanceEnvVar, strconv.Itoa(i)),
```
Comment (arnikola) [inline src/query/server/query.go]: nit: `fmt.Sprintf("%s=%d", multiProcessInstanceEnvVar)`

File: src/query/server/query.go [hunk 8]
```diff
@@ -172,6 +184,109 @@ func Run(runOpts RunOptions) {
...
+					}
+					newEnv = append(newEnv, os.Environ()...)
```
Comment (arnikola) [inline src/query/server/query.go]: nit: prealloc here?

```suggestion
environs := os.Environ()
newEnv := append(make([]string, 0, len(environs) + 1), environs...)
newEnv = append(newEnv, fmt.Sprintf("%s=%d", multiProcessInstanceEnvVar))
```

File: src/query/server/query.go [hunk 9]
```diff
@@ -172,6 +184,109 @@ func Run(runOpts RunOptions) {
...
+
+					opts := panicmon.ExecutorOptions{Env: newEnv}
+					exec := panicmon.NewExecutor(opts)
+					status, err := exec.Run(os.Args)
+					if err != nil {
+						logger.Error("process failed", zap.Error(err))
```
Comment (arnikola) [inline src/query/server/query.go:228]: nit: add status and maybe `multiProcessInstanceEnvVar` to log?

File: src/query/server/query.go [hunk 10]
```diff
@@ -172,6 +184,109 @@ func Run(runOpts RunOptions) {
...
+
+					newEnv := []string{
+						fmt.Sprintf("%s=%s",
+							multiProcessInstanceEnvVar, strconv.Itoa(i)),
+					}
+					newEnv = append(newEnv, os.Environ()...)
+
+					opts := panicmon.ExecutorOptions{Env: newEnv}
+					exec := panicmon.NewExecutor(opts)
+					status, err := exec.Run(os.Args)
```
Comment (arnikola) [inline src/query/server/query.go:226]: Won't this cause issues with e.g. ports?

File: src/query/server/query.go [hunk 11]
```diff
@@ -172,6 +184,109 @@ func Run(runOpts RunOptions) {
...
+				logger.Fatal("could not split host:port for prometheus metrics reporter",
+					zap.Error(err))
+			}
+
+			portValue, err := strconv.Atoi(port)
+			if err != nil {
+				logger.Fatal("prometheus metric reporter port is non-integer",
+					zap.Error(err))
+			}
+			if portValue > 0 {
```
Comment (arnikola) [inline src/query/server/query.go:292]: nit; Add an error case for portValue < 1?

File: src/query/server/query.go [hunk 12]
```diff
@@ -172,6 +184,109 @@ func Run(runOpts RunOptions) {
...
+		// Set the root scope multi-process process ID.
+		if cfg.Metrics.RootScope == nil {
+			cfg.Metrics.RootScope = &instrument.ScopeConfiguration{}
+		}
+		if cfg.Metrics.RootScope.CommonTags == nil {
+			cfg.Metrics.RootScope.CommonTags = make(map[string]string)
+		}
+		cfg.Metrics.RootScope.CommonTags[multiProcessMetricTagID] = multiProcessInstance
+
+		// Listen on a different Prometheus metrics handler listen port.
```
Comment (arnikola) [inline src/query/server/query.go:278]: How does this affect pprof etc endpoints?

File: src/query/server/query.go [hunk 13]
```diff
@@ -172,6 +184,109 @@ func Run(runOpts RunOptions) {
...
+			}
+
+			portValue, err := strconv.Atoi(port)
+			if err != nil {
+				logger.Fatal("prometheus metric reporter port is non-integer",
+					zap.Error(err))
+			}
+			if portValue > 0 {
+				// Increment port value by process ID if valid port.
+				address := net.JoinHostPort(host, strconv.Itoa(portValue+instance))
```
Comment (arnikola) [inline src/query/server/query.go:294]: This may conflict with metrics listen ports

File: src/x/net/listen.go [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+		fn(&result)
+	}
+	return result
+}
+
+// Listen creates a new listener based on options.
+func (o ListenerOptions) Listen(
+	protocol, address string,
+) (gonet.Listener, error) {
+	if o.reusePort {
```
Comment (arnikola) [inline src/x/net/listen.go:68]: nit:
```
if o.reusePort && protocol == "tcp" {
 return reusePortConfig.NewListener(protocol, "tcp4")
}
```

File: src/query/server/query.go (hunk 13 above)
Comment (arnikola) [reply src/query/server/query.go:294]: Oh I see :o

**Pull request merged**